statsmodels/tsa/vecm/vecm.py

```diff
@@ -0,0 +1,704 @@
+    y_min1 : ndarray (neqs x nobs)
+        (dimensions assuming no deterministic terms are given)
+        .. math:: (y_0, \ldots, y_{T-1}
+
+    Returns
+    -------
+    result : tuple
+        A tuple of five ndarrays as well as eigenvalues and -vecotrs of a
+        certain (matrix) product of some of the returned ndarrays.
+    """
+    T = y_min1.shape[1]
+    r0, r1 = _r_matrices(T, delta_x, delta_y_1_T, y_min1)
+    # p. 294: optimizable: e.g. r0.dot(r1.T) == r1.dot(r0.T).T ==> s01==s10.T
+    s00, s01, s10, s11 = (Ri.dot(Rj.T)/T for Ri in (r0, r1) for Rj in (r0, r1))
+    s11_ = inv(mat_sqrt(s11))
+    # p. 295:
+    eig = np.linalg.eig(chain_dot(s11_, s10, inv(s00), s01, s11_))
+    lambd = eig[0]
+    v = eig[1]
+    return s00, s01, s10, s11, s11_, lambd, v
+
+# VECM class: for known or unknown VECM
+
+class VECM(tsbase.TimeSeriesModel):
+    """
+    Fit a VECM process
+    .. math:: \Delta y_t = \Pi y_{t-1} + \Gamma_1 \Delta y_{t-1} + \ldots + \Gamma_{p-1} \Delta y_{t-p+1} + u_t
+    where
+    .. math:: \Pi = \alpha \beta'.
+
+    Parameters
+    ----------
+    endog_tot : array-like
+        2-d endogenous response variable.
+    dates : array-like
+        must match number of rows of endog
+    method : {"ls", "egls", "ml"}
+        Estimation method to use.
+    diff_lags : int
+        Number of lags in the VEC representation
+    deterministic : str {"nc", "co", "ci", "lo", "li"}
+            "nc" - no deterministic terms
+            "co" - constant outside the cointegration relation
+            "ci" - constant within the cointegration relation
+            "lo" - linear trend outside the cointegration relation
+            "li" - linear trend within the cointegration relation
+            Combinations of these are possible (e.g. "cili" or "colo" for
+            linear trend with intercept)
+    seasons : int
+        Number of seasons. 0 (default) means no seasons.
+    coint_rank : int
+        Cointegration rank, equals the rank of the matrix \Pi and the number
+        of columns of \alpha and \beta
+
+    References
+    ----------
+    [Lut2005]_
+    .. [Lut2005] Lutkepohl (2005) New Introduction to Multiple Time Series Analysis
+    """
+    
+    # TODO: implementation
+    def __init__(self, endog_tot, dates=None, freq=None, missing="none",
+                 method="ml", diff_lags=1, deterministic="nc", seasons=0,
+                 coint_rank=1):
+        super(VECM, self).__init__(endog_tot, None, dates, freq,
+                                   missing=missing)
+        if self.endog.ndim == 1:
+            raise ValueError("Only gave one variable to VECM")
+        self.y = self.endog.T  # TODO delete this line if y not necessary
+        self.neqs = self.endog.shape[1]
+        self.p = diff_lags + 1
+        self.method = method
+        self.diff_lags = diff_lags
+        self.deterministic = deterministic
+        self.seasons = seasons
+        self.coint_rank = coint_rank
+
+    def fit(self):
```

Comment:
Things that don't change directly the underlying data/regressor arrays can stay in `fit`
`method` is for the fit and is a `fit` argument in all models
`coint_rank`  only affects the parameters and not the underlying regressors. so, I think it should be safe to keep it in fit.